### PR TITLE
Initial tests

### PR DESCRIPTION
**Pull request by DhairyaLGandhi**
ref #11

### CODE REVIEW

**Comment by DhairyaLGandhi** [comment]
I have tried to use some more descriptive names in the `check_data_parallel` function so hopefully it shouldn’t be too confusing to follow along.

The base case we are trying to assert is that the gradients that came out of a batch of data is the sum of the gradients from the individual data points in the batch. We can load the same images as well, but that’s another layer of complication(#) that I want to test explicitly. For the current case of `check_data_parallel` , we are therefore using a `rand` tensor (can be made a fill!, ones, … to ensure constant data going through the model).

So now, we are trying to compare grads from a batch to the gradients from individual datapoints in the batch. Coming to `check_data_parallel`:

1. Move data and model to GPU - `gpu_data, gpu_model`
2. Copy the model before any updates to have a clean instance of the model(*). This is `gpu_model2`
3. `batchedgrads` are the gradients coming out of a regular batch of data - our ground truth.
4. Get the distributed gradients (`distributedgrads_x3` (^)):
         a. iterate over the batch and get a hold of the data points (`gpu_data[colons...,i:i]`)
         b. copy the clean model (`gpu_model2`; same reason as *)
         c. get the grads for the datapoint
         d. This will yield a vector of gradient objects corresponding to every datapoint in the batch
5. With `distributedgrads_x3`, we need to add all the gradient objects in this vector. We do it in 2 ways:
          a. `final` - `Functor.fmap(…) do _accum(…) end` This is how it is done in the package, and we want to check we are doing this correctly
          b. Manually adding arrays to remove the complexity of 5a to get a ground truth to test 5a against. This is done via the `get_sum` function. `getfirst(grad, field)` iterates over tuples, looking for namedtuples that have field and returning the first match.
6. (Not done in the tests, but I had checked locally anyway). Check 5a == 5b
7. Compare($) 5a (`final`) with ground truth (`batchedgrads`)
8. Compare 5b (`get_sum(getfirst.(distributedgrads_x3, :weight)`) with `getfirst(batchedgrads, :weight)`

i. Run this testing scheme on individual layers (`Dense, Conv, MaxPool`, …)
ii. Run it for simple Chains to test layer composition and gradients
iii. Run it for layers with effects (`BatchNorm`)
iv. Run it for BatchNorm with other layers (Chains with `BatchNorm` in them, the example in the test suite runs `Chain(Conv((7,7), 3 => 3), BatchNorm(3))`. This is because it is a common pattern in conv models and resnets)
Running the test scheme shows i., ii., iii. all pass.

iv. passes for models that don't have effects - ie the `BatchNorm` is frozen. This is expected, most frameworks do things differently from here on. TF has a `SyncBatchNorm` which is an experimental layer that syncs the running stats during training. Torch doesn't sync stats, since with long enough training, all these stats should converge.

#: whether the images are loaded and batched correctly given the parallel nature of our batching - not a big deal, just something that needs to be asserted in tests
*: We don’t have GPUs on GA, and only a single GPU on BK, so in order to simulate multiple instances of the model for DDP, get a copy of the model before any side effects such as running stats from BatchNorm
^: x3 is because we have 3 datapoints in the batch by default, so it is a vector with 3 gradient objects, each corresponding to every data point in the batch.
$: `compare(a, b)` walks over fields of identical structs (`compare(a::T, b::T) where T`), tuples and fields of NamedTuples and runs `\approx` with the leaves of the trees - these are our parameters from our model.

The gradient synchronisation works as expected, which is excellent. The current breakage is due to `sync_buffer` using a `CUDA.device!` which errors when no GPU is present. I am setting up the groundwork to ensure optimisation is tested, as well as the training loop is tested in a manner representative of the inner workings of the package. This seemed to be far more important than the data loading to make sure our tests cover the most used bits very strongly.

**Comment by DhairyaLGandhi** [comment]
I have also added the optimisation tests to this pr, which should help cover a lot of the pieces that this project interacts with.
